File: python system/inventory_system/services/audit_logger.py
import json
import uuid
from typing import Optional
from datetime import datetime
from models.audit_log import AuditLog, AuditAction
from models.inventory_item import InventoryItem
from repositories.audit_log_repository import AuditLogRepository


class AuditLogger:
    def __init__(self, repo: AuditLogRepository):
        self._repo = repo

    def log_item_created(self, item: InventoryItem, user_id: Optional[str] = None) -> None:
        self._log(
            action=AuditAction.CREATED,
            entity_type="InventoryItem",
            entity_id=item.id,
            user_id=user_id,
            before_snapshot=None,
            after_snapshot=json.dumps(item.to_dict()),
        )

    def log_item_updated(
        self,
        before: InventoryItem,
        after: InventoryItem,
        user_id: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> None:
        self._log(
            action=AuditAction.UPDATED,
            entity_type="InventoryItem",
            entity_id=after.id,
            user_id=user_id,
            before_snapshot=json.dumps(before.to_dict()),
            after_snapshot=json.dumps(after.to_dict()),
            notes=notes,
        )

    def log_item_deleted(self, item: InventoryItem, user_id: Optional[str] = None) -> None:
        self._log(
            action=AuditAction.DELETED,
            entity_type="InventoryItem",
            entity_id=item.id,
            user_id=user_id,
            before_snapshot=json.dumps(item.to_dict()),
            after_snapshot=None,
        )

    def log_stock_in(
        self,
        before: InventoryItem,
        after: InventoryItem,
        user_id: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> None:
        self._log(
            action=AuditAction.STOCK_IN,
            entity_type="InventoryItem",
            entity_id=after.id,
            user_id=user_id,
            before_snapshot=json.dumps(before.to_dict()),
            after_snapshot=json.dumps(after.to_dict()),
            notes=notes,
        )

    def log_stock_out(
        self,
        before: InventoryItem,
        after: InventoryItem,
        user_id: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> None:
        self._log(
            action=AuditAction.STOCK_OUT,
            entity_type="InventoryItem",
            entity_id=after.id,
            user_id=user_id,
            before_snapshot=json.dumps(before.to_dict()),
            after_snapshot=json.dumps(after.to_dict()),
            notes=notes,
        )

    def log_adjustment(
        self,
        before: InventoryItem,
        after: InventoryItem,
        user_id: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> None:
        self._log(
            action=AuditAction.ADJUSTMENT,
            entity_type="InventoryItem",
            entity_id=after.id,
            user_id=user_id,
            before_snapshot=json.dumps(before.to_dict()),
            after_snapshot=json.dumps(after.to_dict()),
            notes=notes,
        )

    def _log(
        self,
        action: AuditAction,
        entity_type: str,
        entity_id: str,
        user_id: Optional[str] = None,
        before_snapshot: Optional[str] = None,
        after_snapshot: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> None:
        log = AuditLog(
            id=str(uuid.uuid4()),
            timestamp=datetime.now(),
            action=action,
            entity_type=entity_type,
            entity_id=entity_id,
            user_id=user_id,
            before_snapshot=before_snapshot,
            after_snapshot=after_snapshot,
            notes=notes,
        )
        self._repo.save(log)

File: dart/transaction_system/dart_application_1/bin/audit_logger.dart
// lib/services/audit_logger.dart

import 'dart:convert';
import 'package:uuid/uuid.dart';

import 'models/audit_log.dart';
import 'models/inventory_item.dart';
import 'audit_log_repository.dart';

class AuditLogger {
  final AuditLogRepository _repo;
  final Uuid _uuid;

  AuditLogger(this._repo) : _uuid = const Uuid();

  void logItemCreated(InventoryItem item, {String? userId}) {
    _log(
      action: AuditAction.created,
      entityType: 'InventoryItem',
      entityId: item.id,
      userId: userId,
      beforeSnapshot: null,
      afterSnapshot: jsonEncode(item.toJson()),
    );
  }

  void logItemUpdated(
    InventoryItem before,
    InventoryItem after, {
    String? userId,
    String? notes,
  }) {
    _log(
      action: AuditAction.updated,
      entityType: 'InventoryItem',
      entityId: after.id,
      userId: userId,
      beforeSnapshot: jsonEncode(before.toJson()),
      afterSnapshot: jsonEncode(after.toJson()),
      notes: notes,
    );
  }

  void logItemDeleted(InventoryItem item, {String? userId}) {
    _log(
      action: AuditAction.deleted,
      entityType: 'InventoryItem',
      entityId: item.id,
      userId: userId,
      beforeSnapshot: jsonEncode(item.toJson()),
      afterSnapshot: null,
    );
  }

  void logStockIn(
    InventoryItem before,
    InventoryItem after, {
    String? userId,
    String? notes,
  }) {
    _log(
      action: AuditAction.stockIn,
      entityType: 'InventoryItem',
      entityId: after.id,
      userId: userId,
      beforeSnapshot: jsonEncode(before.toJson()),
      afterSnapshot: jsonEncode(after.toJson()),
      notes: notes,
    );
  }

  void logStockOut(
    InventoryItem before,
    InventoryItem after, {
    String? userId,
    String? notes,
  }) {
    _log(
      action: AuditAction.stockOut,
      entityType: 'InventoryItem',
      entityId: after.id,
      userId: userId,
      beforeSnapshot: jsonEncode(before.toJson()),
      afterSnapshot: jsonEncode(after.toJson()),
      notes: notes,
    );
  }

  void logAdjustment(
    InventoryItem before,
    InventoryItem after, {
    String? userId,
    String? notes,
  }) {
    _log(
      action: AuditAction.adjustment,
      entityType: 'InventoryItem',
      entityId: after.id,
      userId: userId,
      beforeSnapshot: jsonEncode(before.toJson()),
      afterSnapshot: jsonEncode(after.toJson()),
      notes: notes,
    );
  }

  void _log({
    required AuditAction action,
    required String entityType,
    required String entityId,
    String? userId,
    String? beforeSnapshot,
    String? afterSnapshot,
    String? notes,
  }) {
    final log = AuditLog(
      id: _uuid.v4(),
      timestamp: DateTime.now(),
      action: action,
      entityType: entityType,
      entityId: entityId,
      userId: userId,
      beforeSnapshot: beforeSnapshot,
      afterSnapshot: afterSnapshot,
      notes: notes,
    );
    _repo.save(log);
  }
}
